ServiceTier/Services/LogInService.cs

```diff
@@ -33,39 +36,60 @@ public string CreateToken(LoginDTO model)
 
-            if(!userManager.UserManager.CheckPasswordAsync(user, model.Password).Result)
+            if (!userManager.UserManager.CheckPasswordAsync(user, model.Password).Result)
             {
                 throw new IncorectLoginException();
             }
 
-            var token = GenerateToken(user);
+            var token = JWTManager.GenerateToken(user);
 
             return token;
         }
 
-        private static string GenerateToken(User user)
+        public string CreateTokenWithGoogle(SocialAuthDTO accessToken)
         {
-            var claims = new[]
-            {
-                new Claim(Constants.ClaimUserId, user.Id.ToString()),
-                new Claim(JwtRegisteredClaimNames.Sub, user.Email),
-                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
-            };
+            string userInfoResponse = Client.GetStringAsync($"{Constants.GoogleResponsePath}{accessToken.AccessToken}").Result;
+
+            var userInfo = JsonConvert.DeserializeObject<GoogleAuthUserData>(userInfoResponse);
+            User user = userManager.UserManager.FindByEmailAsync(userInfo.Email).Result;
 
-            var signingKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(IoCContainer.Configuration["JWTSecretKey"]));
+            if (user == null)
+            {
+                user = userInfo.ToUser();
+                var result = userManager.UserManager.CreateAsync(user, Convert.ToBase64String(Guid.NewGuid().ToByteArray()).Substring(0, 8)).Result;
 
-            var token = new JwtSecurityToken(
-                expires: DateTime.UtcNow.AddHours(Constants.Validity),
-                claims: claims,
-                signingCredentials: new SigningCredentials(signingKey, SecurityAlgorithms.HmacSha256),
-                issuer: "Tokens:Issuer",
-                audience: "Tokens:Audience"
-                );
+                if (!result.Succeeded)
+                {
+                    throw new UserNotAddedException();
+                }
+            }
 
-            var tokenString = new JwtSecurityTokenHandler().WriteToken(token);
+            string jwt = JWTManager.GenerateToken(user);
 
-            return JsonConvert.SerializeObject(new { access_token = tokenString });
+            return jwt;
         }
 
+        public string CreateTokenWithFacebook(SocialAuthDTO accessToken)
+        {
+            string userInfoResponse = Client.GetStringAsync($"{Constants.FacebookResponsePath}{accessToken.AccessToken}").Result;
+
+            var userInfo = JsonConvert.DeserializeObject<FacebookAuthUserData>(userInfoResponse);
+            User user = userManager.UserManager.FindByEmailAsync(userInfo.Email).Result;
+
+            if (user == null)
+            {
+                user = userInfo.ToUser();
+                var result = userManager.UserManager.CreateAsync(user, Convert.ToBase64String(Guid.NewGuid().ToByteArray()).Substring(0, 8)).Result;
+
+                if (!result.Succeeded)
```

Comment:
It is an additional check for the success of the operation of adding a new user. If it method fails we don't invoke method for generation token.

Reply:
I wasn't asking about the check, check is fine. I was wondering if we need brand new UserNotAddedException. Because as I understand it - this is the only place where such exception will be used, right?

Reply:
Yes, this exception is used only in this place. Maybe use general exception and pass information in message parameter, for example, new Exception("Didn`t add new user")?

Reply:
yes, that is a good solution, and the error message should be something like : 'couldn't add / create new user'

Reply:
ok